src/log.rs

```diff
@@ -1,47 +1,152 @@
-use crate::include::common::validate::validate_input;
-use crate::include::dav1d::dav1d::Rav1dLogger;
-use crate::src::internal::Rav1dContext;
-use crate::stderr;
+use cstr::cstr;
 use std::ffi::c_char;
-use std::ffi::c_int;
+use std::ffi::c_uint;
 use std::ffi::c_void;
+use std::fmt;
+use std::fmt::Write as _;
+use std::io::stderr;
+use std::io::stdout;
+use std::io::Write as _;
 use std::ptr;
 
-extern "C" {
-    fn vfprintf(_: *mut libc::FILE, _: *const c_char, _: ::core::ffi::VaList) -> c_int;
+pub type Dav1dLoggerCallback = unsafe extern "C" fn(
+    // The above `cookie` field.
+    cookie: *mut c_void,
+    // A `printf`-style format specifier.
+    fmt: *const c_char,
+    // `printf`-style variadic args.
+    args: ...
+);
+
+#[derive(Clone)]
+#[repr(C)]
+pub struct Dav1dLogger {
+    /// A cookie that's passed as the first argument to the callback below.
+    cookie: *mut c_void,
+    /// A `printf`-style function except for an extra first argument that will always be the above `cookie`.
+    callback: Option<Dav1dLoggerCallback>,
 }
 
-#[cold]
-pub unsafe extern "C" fn rav1d_log_default_callback(
-    _cookie: *mut c_void,
-    format: *const c_char,
-    mut ap: ::core::ffi::VaList,
-) {
-    vfprintf(stderr, format, ap.as_va_list());
+impl Dav1dLogger {
+    /// # Safety
+    ///
+    /// `callback`, if non-[`None`]/`NULL` must be safe to call when:
+    /// * the first argument is `cookie`
+    /// * the rest of the arguments would be safe to call `printf` with
+    pub const unsafe fn new(cookie: *mut c_void, callback: Option<Dav1dLoggerCallback>) -> Self {
+        Self { cookie, callback }
+    }
 }
 
-impl Default for Rav1dLogger {
+impl Default for Dav1dLogger {
     fn default() -> Self {
-        Self {
-            cookie: ptr::null_mut(),
-            callback: Some(rav1d_log_default_callback),
+        // Safety: `callback` is [`None`].
+        unsafe { Self::new(ptr::null_mut(), None) }
+    }
+}
+
+/// A [`Dav1dLogger`] from C that's not
+/// just a [`Rav1dLogger`] converted to a [`Dav1dLogger`].
+#[derive(Clone)]
+pub(crate) struct OnlyDav1dLogger {
+    cookie: *mut c_void,
+    callback: Dav1dLoggerCallback,
+}
+
+impl fmt::Write for OnlyDav1dLogger {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        // `s` doesn't have a terminating nul-byte,
+        // and it may have internal nul-bytes,
+        // so it's easiest just to print one byte at a time.
+        // This may be slow, but logging can be disabled if it's slow,
+        // or the Rust API can be used instead.
+        let fmt = cstr!("%c");
```

Comment:
Native `c"strings"` were just stabilized in nightly so we'll switch to that once it lands in a couple months.  I think the crate is fine in the meantime as it has an identical API.

Reply:
We can switch to `c""` strings only once we know that potential downstream users are on a recent enough Rust version. I'm not sure I understand what you mean by an identical API, the new literal syntax doesn't use a macro.

That all said, that has no bearing on whether we should add a new dependency just for a single line of convenience. A single line is not enough to justify a new dependency, regardles of how short-lived you expect that to be. Once you add something it often stays longer than you'd think. If the crate was providing something we couldn't have otherwise, then fine, but it's not here. Don't use it.

Reply:
Isn't `cstr!` the canonical way to write C string literals (before the native way is stabilized)?  I'm just trying to stick to the easiest, simplest way.  I don't think we should worry about these little things but focus more of making things safe fast.  Choices like this are irrelevant until we do and we can revisit things like this later.  I'm fine changing this to `CStr::from_bytes_with_nul(b"%c\0")`, but I don't think we should care either way until everything's safe.

Reply:
I only care here because it adds a new dependency. That's enough of a reason to not use it unless we plan to make _lots_ of these C literals, and I don't think we need to?

Reply:
> and I don't think we need to?

Probably not?  I didn't look enough to see for sure.

I'll replace this with `CStr::from_bytes_with_nul(b"%c\0").unwrap()` and then `c"%c"` when its stabilization lands on stable.  As for not updating until we know users have a recent enough rust version, that's pretty hard, as we're already on a recent `nightly` and aren't paying any attention to any MSRV things.

Reply:
Changed in 9d47278.